### PR TITLE
Component enum resolving in local compilation mode

### PR DESCRIPTION
**Pull request by pmvald**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts [hunk 1]
```diff
@@ -34,6 +35,32 @@ export function resolveEnumValue(
   return resolved;
 }
 
+/** Resolves a EncapsulationEnum expression locally on best effort. */
+export function resolveEncapsulationEnumValueLocally(expr?: ts.Expression): number|null {
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts:39]: isn't it possible to still try the static interpretation? especially if e.g. a variable is used from the same file?

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts:39]: We should also include in the JSDoc that we need a static value at compile time for e.g. styles processing etc.

File: packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts [hunk 2]
```diff
@@ -34,6 +35,32 @@ export function resolveEnumValue(
...
+  if (!expr) {
+    return null;
+  }
+
+  const exprText = expr.getText().trim();
+
+  for (const key in ViewEncapsulation) {
+    if (Number.isFinite(Number(key))) {
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts:47]: Nit: is `isNaN` more clear here? (add a comment to filter our enum values?)

**Review by devversion — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts (hunk 1 above)
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts:45]: No this enum is always (and should be) imported from `@angular/core`. It cannot be in the same file.

JsDoc added

File: packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts (hunk 2 above)
Comment (pmvald) [reply packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts]: Done!

File: packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts (hunk 1 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/common/src/evaluation.ts:45]: Oh I meant, if someone does something like:

```ts
const encapsulation = ViewEncapsulation.X

@Component({
  encapsulation,
})
```

but I can see how that is difficult and highly unlikely. The JSDoc was the important part. Thx

**Review by devversion — APPROVED** [review]

**Comment by pmvald** [comment]
Caretaker note: the google presubmit failures seem to be already failing

**Review by crisbeto — APPROVED** [review]

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit f91f222b55f249089d267c72a9c0ab5b09d7c932.

**Pull request closed without merging**